runtime/src/sorted_storages.rs

```diff
@@ -120,6 +134,66 @@ impl<'a> SortedStorages<'a> {
     }
 }
 
+/// Iterator over successive slots in 'storages' within 'range'.
+/// This enforces sequential access so that random access does not have to be implemented.
+/// Random access could be expensive with large sparse sets.
+pub struct SortedStoragesIter<'a> {
+    /// range for the iterator to iterate over (start_inclusive..end_exclusive)
+    range: Range<Slot>,
+    /// the data to return per slot
+    storages: &'a SortedStorages<'a>,
+    /// the slot to be returned the next time 'Next' is called
+    next_slot: Slot,
+}
+
+impl<'a> Iterator for SortedStoragesIter<'a> {
+    type Item = (Slot, Option<&'a SnapshotStorage>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let slot = self.next_slot;
+        if slot < self.range.end {
+            // iterator is still in range. Storage may or may not exist at this slot, but the iterator still needs to return the slot
+            self.next_slot += 1;
+            Some((slot, self.storages.get(slot)))
+        } else {
+            // iterator passed the end of the range, so from now on it returns None
+            None
+        }
+    }
+}
+
+impl<'a> SortedStoragesIter<'a> {
+    pub fn new<R: RangeBounds<Slot>>(
+        storages: &'a SortedStorages<'a>,
+        range: R,
+    ) -> SortedStoragesIter<'a> {
+        let storage_range = storages.range();
+        let next_slot = match range.start_bound() {
+            Bound::Unbounded => {
+                storage_range.start // unbounded beginning starts with the min known slot (which is inclusive)
+            }
+            Bound::Included(x) => *x,
+            Bound::Excluded(x) => *x + 1, // make inclusive
+        };
+        let end_exclusive_slot = match range.end_bound() {
+            Bound::Unbounded => {
+                storage_range.end // unbounded end ends with the max known slot (which is exclusive)
+            }
+            Bound::Included(x) => *x + 1, // make exclusive
+            Bound::Excluded(x) => *x,
+        };
```

Comment:
Could the bounds be used directly here, and then just check-and-fixup iff unbounded? In other words, can you construct the range from `Bounds` objects instead of the slot numbers directly? I was thinking this might be a bit simpler and thus avoid any later off-by-one issues.

Reply:
I have found bounds to be cumbersome to work with. I wanted to convert into a fixed inclusive..exclusive range, using the info we have from the storages. We can refactor this isolated part later if we can come up with a better implementation.